Add features needed by `patch` library

endgame opened the pull request:
See also: https://github.com/reflex-frp/patch/issues/4

This PR:

* Adds `(~~)` as a group subtraction operator, as it was in `patch`'s
  `Group` class. It is `infixl 7` based on the discussion in
  https://github.com/reflex-frp/patch/issues/4 .

* Adds instances for every type present in `base`. Instances for
  `MonoidalMap` are not ported across, because `monoidal-containers`
  pulls in `lens`. If this PR is accepted, I will PR
  `monoidal-containers` to move those instances.

* Adds a few obvious `Cyclic` instances.

Note: To write instances for `(:*:)` and `(:.:)`, I had to turn on
`-XTypeOperators`. This means that we're no longer pure Haskell98. The
cabal documentation has been updated to reflect this. I wrote out the
instances for `Const` and `Identity` by hand to avoid turning on
`-XStandaloneDeriving` and `-XGeneralizedNewtypeDeriving`.

endgame committed:
Whitespace cleanup

endgame committed:
Add group subtraction operator

endgame committed:
Port over instances defined in https://github.com/reflex-frp/patch

endgame committed:
Add some obvious Cyclic instances

endgame commented on groups.cabal:
```diff
@@ -14,6 +14,6 @@ cabal-version:       >=1.8
-  -- other-modules:       
-  build-depends:       base <5
+  -- other-modules:
+  build-depends:       base >= 4.12.0.0 && < 5
```
The lower bound is when the `Monoid` instances for `(:*:)` and `(:.:)` appeared.

Taneb replied:
Hmm, I think I'd rather we used CPP so we could still support at least as far back as 7.6 (the version of GHC in debian oldoldstable and CentOS 7). I'm going to finally add a test suite soon so we can check this automatically!

endgame replied:
New base bound supports GHC 7.6.1 and above, but creates a lot of CPP noise. `{-# LANGUAGE TypeOperators #-}` has been around since GHC 6.8.1 so I didn't try to CPP that away.

endgame committed:
CPP away some of the instances under old GHC and maintain compat

Use `mappend` to support versions of base that doen't have Semigroup.

endgame committed:
Also guard the Cyclic instances behind CPP

Taneb commented on src/Data/Group.hs:
```diff
@@ -101,3 +123,62 @@ generated :: Cyclic a => [a]
+-- | Trivial group, Functor style.
+instance Group (Proxy x) where
+  invert _ = Proxy
+  _ ~~ _ = Proxy
```
I think an explicit definition of `pow` is worth it here and for other instances that are isomorphic to `()`

endgame replied:
Done. I think `Proxy` is the only such type?

endgame committed:
instance Group (Proxy x): Add explicit implementation of pow

endgame committed:
Fix reference in LICENSE to sources from `patch`

endgame committed:
Fix haddock

endgame committed:
Comment nit

endgame commented:
Is there anything else that needs to happen here?

endgame commented:
Just checking in again - is there anything else you need from me for this?

Taneb commented on src/Data/Group.hs:
```diff
@@ -1,14 +1,32 @@
+{-# LANGUAGE CPP           #-}
+{-# LANGUAGE TypeOperators #-}
```
Is `TypeOperators` only needed for the `GHC.Generics` instances? If that's the case, could it be in a CPP conditional block (with a comment to explain why)

endgame replied:
Yes, but `-XTypeOperators` landed in 6.8.1 (the `GHC.Generics` types didn't show up until later, which is why their CPP is for GHC 7.x). 6.8.1 is older than oldstable, by your previous remarks in this PR, so is more CPP necessary here? I will happily add it if you insist on it (e.g., you want to preserve Haskell98 compatibility for as long as possible), but just want to check first.

Taneb replied:
I'd rather you add it, please

endgame replied:
No problem at all. I should be able to push it up later this week.

endgame replied:
Done. Anything else I can do for you?

endgame committed:
CPP away -XTypeOperators when not needed

endgame commented:
@Taneb Anything else you need from me here?

endgame commented:
@Taneb Anything else necessary to get this merged?

Taneb commented:
Having thought it over for a while, I'm not happy with the splitting of the copyright for this library. Either can you transfer me the copyright, or I can ask someone who hasn't seen this PR to implement those particular instances from scratch.

endgame commented:
It's not my copyright to transfer. I can ping someone from Obsidian but I think it would be faster to ask someone to cleanroom it.

ryantrinkle commented:
@Taneb I'm a partner at Obsidian.  We're fine with relicensing this under your current license, which appears to be exactly the same as ours except for a minor change of wording and formatting.

@endgame Thanks for taking the time to integrate this here!

endgame committed:
Remove Obsidian Systems from LICENSE

Ryan Trinkle (Partner, Obsidian Systems) consented to this:
https://github.com/taneb/groups/pull/3#issuecomment-670782175

endgame commented:
Okay @Taneb, how's this?

endgame commented:
Thanks heaps!

Taneb commented:
I'll get this released ASAP

Taneb commented:
And released! https://hackage.haskell.org/package/groups-0.5
